internal/mempool/mempool.go

```diff
@@ -872,19 +876,154 @@ func (g *txDescGraph) forEachDescendant(txHash *chainhash.Hash,
+// pre-order.
+// If the provided function f returns true then the traversal will continue
+// to walk descendants of the respective transaction.
+func (g *txDescGraph) forEachDescendantPreOrder(txHash *chainhash.Hash,
+	seen map[chainhash.Hash]struct{}, f func(*mining.TxDesc) bool) {
+	for child, childDesc := range g.childrenOf[*txHash] {
+		if _, saw := seen[child]; saw {
+			continue
+		}
+
+		seen[child] = struct{}{}
+		if f(childDesc) {
+			g.forEachDescendantPreOrder(&child, seen, f)
+		}
+	}
+}
+
+// forEachAncestorPostOrder iterates over all transactions in the graph that
+// txHash depends on and invokes the function f for each transaction,
+// in post-order.
+// If the provided function f returns true then the traversal will continue
+// to walk ancestors of the respective transaction.
+func (g *txDescGraph) forEachAncestorPostOrder(txHash *chainhash.Hash,
+	seen map[chainhash.Hash]struct{}, f func(tx *mining.TxDesc) bool) {
+
+	for parent, parentDesc := range g.parentsOf[*txHash] {
+		if _, saw := seen[parent]; saw {
+			continue
+		}
+
+		seen[parent] = struct{}{}
+		if f(parentDesc) {
+			g.forEachAncestorPostOrder(&parent, seen, f)
+		}
+	}
+}
+
+// maybeUpdateAncestorStats attempts to update the ancestor stats for a given
+// transaction. It walks the graph for all ancestors of the provided transaction
+// and aggregates statistics for its ancestors.
+func (mv *txMiningView) maybeUpdateAncestorStats(txDesc *mining.TxDesc) bool {
+	txHash := txDesc.Tx.Hash()
+	canTrackAncestors := true
+	stats := &mining.TxAncestorStats{}
+	seenAncestors := make(map[chainhash.Hash]struct{}, ancestorTrackingLimit+1)
+	mv.txGraph.forEachAncestorPostOrder(txHash, seenAncestors,
+		func(ancestorTxDesc *mining.TxDesc) bool {
+			if !canTrackAncestors {
+				// Short circuit the below checks if we know that the provided
+				// transaction cannot have ancestor stats aggregated.
+				return false
+			}
+
+			if len(seenAncestors) > ancestorTrackingLimit {
+				// A transaction with too many ancestors should not have
+				// ancestor stats cached.
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor transaction does not have stats, then the
+			// provided one should not either.
+			ancestorTxHash := *ancestorTxDesc.Tx.Hash()
+			ancestorStats, exists := mv.ancestorStats[ancestorTxHash]
+			if !exists {
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor plus its respective ancestors would exceed the
+			// limit for this transaction, then do not track stats.
+			if ancestorStats.NumAncestors+1 > ancestorTrackingLimit {
+				canTrackAncestors = false
+				return false
+			}
+
+			// If any ancestor has more children than the tracking limit,
+			// then stats should not be tracked for the provided transaction.
+			ancestorChildren := mv.txGraph.childrenOf[ancestorTxHash]
+			if len(ancestorChildren) >= ancestorTrackingLimit {
```

Comment:
I think it would be beneficial to modify the template generator to short circuit anything with ancestor statistics tracked if too many transactions have been rejected during block template generation. Doing that would reduce the amount of churn the block template generator runs into when encountering a chain that maximizes the number of tracked descendants restricted only by the number of children and minimum mempool fee.

An even more restrictive policy would limit by the number of transactions reachable from the base transaction `txDesc`, and would only require a single additional allocation for a reachable transaction map like

```suggestion
			mv.txGraph.forEachDescendantPreOrder(ancestorTxHash, reachableTxns,
				func(ancestorDescendantTxDesc *mining.TxDesc) bool {
					if !canTrackAncestors {
						return false
					}
					if len(reachableTxns) > reachableTransactionLimit {
						canTrackAncestors = false
						return false
					}
					return true
				})
```

For prioritizing performance in the short-term, I'd opt for limiting by reachability though I'd prefer less restrictive behavior in the long-term. Does anyone else have thoughts to share on this scenario?